src/ch/blinkenlights/android/vanilla/MediaUtils.java

```diff
@@ -426,12 +428,36 @@ public static Song getRandomSong(Context context, boolean albumShuffle) {
 			sSongCount = songs.size();
 		}
 
-		Song result = null;
+		final List<Song> results = new ArrayList<>();
+
 		if (songs.size() > 0) {
-			result = songs.remove(0);
-			result.flags |= Song.FLAG_RANDOM;
+			final Song song = songs.remove(0);
+
+			if (!albumShuffle) {
+				song.flags |= Song.FLAG_RANDOM;
+			}
+
+			results.add(song);
+
+			// since we're in album shuffle mode, we'll want to add in the entire album in one go,
+			// so loop through the upcoming songs and add all those that have the same album id
+			// as the song we initially got
+			if (songs.size() > 0 && albumShuffle) {
+				long expectedAlbumId = song.albumId;
+				boolean hasMore = true;
+
+				while (hasMore) {
+					hasMore = false;
+					if (songs.size() > 0 && songs.get(0).albumId == expectedAlbumId) {
```

Comment:
This looks clunky :-)

How about a `do {} while()` loop?

Reply:
Yep. Changed to a do...while{} loop.